### PR TITLE
Add Discontinuous Galerkin capabilities to the Tracer auxiliary physics

### PR DESCRIPTION
**Pull request by blaisb**
### Description

Adds the capability to the Tracer to use a Discontinuous Galerkin formulation instead of a Continuous Galerkin formulation. When DG is used, an entirely different paradigm has to be used to calculate the integral at the internal faces and at the boundary faces. This PR deploys this architecture as well as numerous application_test to test the DG implementation.

Concretely, here is a summary of what is done:
- A new assembly routine for the RHS and the Matrix have been added to the tracer. This new assembly routine uses the MeshWorker paradigm to loop over inner faces, boundary faces and cells. 
- The tracer scratch data has been modified to accomodate the necessary DG features. This includes calculation of the velocity at the internal/boundary faces, calculation of the tracer terms at the faces and calculation of the physical properties at the faces.
- An outlet boundary condition has been added for the tracer physics.
- Assemblers for the DG equations have been added.

### Testing

9 additional application_tests were introduce to fully test the DG implementation.

### Documentation

Few modifications are required in the .RST files, they have been added.

### Miscellaneous (will be removed when merged)

This is a feature I have been working on for a long time.  The results are very encouraging. There will be some slight rework to be done in the future, but first I will have to rearchitecture how boundary conditions are calculated.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [X] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [X] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [X] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [X] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [X] Labels are applied
- [X] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [X] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [X] If the fix is temporary, an issue is opened
- [X] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/solvers/tracer_scratch_data.h [hunk 1]
```diff
@@ -194,8 +213,12 @@ class TracerScratchData
             this->laplacian_phi[q][k] = trace(this->hess_phi[q][k]);
           }
       }
+
+    boundary_index = 0;
   }
 
+
```
Comment (mivaia) [inline include/solvers/tracer_scratch_data.h:220]: extra lines

File: include/solvers/tracer.h [hunk 2]
```diff
@@ -383,6 +423,120 @@ class Tracer : public AuxiliaryPhysics<dim, GlobalVectorType>
...
+   *
+   * @param[in] face_no the number of the face where the velocity is to be
+   * reinitialized.
+   *
+   * @param[in,out] scratch_data the scratch data to be used for the
+   * reinitialization.
+   */
+  inline void
+  reinit_face_velocity_with_adequate_solution(
+    const typename DoFHandler<dim>::active_cell_iterator &velocity_cell,
```
Comment (mivaia) [inline include/solvers/tracer.h:476]: I know this is present in other parts of the code in order to use the average velocity or not. Is there an easy way to get rid of sections like this one scattered in the code? I feel it takes a lot of place and could simplify some sections.

**Review by mivaia — CHANGES_REQUESTED** [review]
Really nice work! Made some preliminary comments. I will look at the tests when I have some time tomorrow.

File: include/core/boundary_conditions.h [hunk 3]
```diff
@@ -253,19 +254,17 @@ namespace BoundaryConditions
...
     prm.leave_subsection();
 
-    // Penalization parameter for weakly imposed dirichlet BCs and outlets
     prm.declare_entry(
       "beta",
-      "0",
+      "1",
       Patterns::Double(),
-      "penalty parameter for weak boundary condition imposed through Nitsche's method or outlets");
+      "Penalty parameter for weak boundary condition imposed through Nitsche's method or outlets");
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:261]: The beta coefficient for regular Nitsche method or outlet is dependant (in the case of the Navier-Stokes equations). in the case of the Tracer, This parameter was actually not needed.

File: source/solvers/tracer.cc [hunk 4]
```diff
@@ -80,6 +108,147 @@ Tracer<dim>::assemble_system_matrix()
...
+                                                  face_no,
+                                                  scratch_data);
+
+      scratch_data.calculate_face_physical_properties();
+
+      this->inner_face_assembler->assemble_matrix(scratch_data, copy_data);
+    };
+
+
+  const auto copier = [&](const StabilizedDGMethodsCopyData &c) {
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer.cc:223]: Copier here as in the thing that copies ;)

File: include/solvers/tracer_scratch_data.h [hunk 5]
```diff
@@ -260,6 +284,219 @@ class TracerScratchData
...
+    const double extent1 = cell->measure() / cell->face(face_no)->measure();
+    const double extent2 = ncell->measure() / ncell->face(nf)->measure();
+
+    penalty_factor =
+      get_penalty_factor(fe_values_tracer.get_fe().degree, extent1, extent2);
+
+
+    // BB TODO : Preallocate memory here
+    values_here.resize(face_quadrature_points.size());
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_scratch_data.h:335]: Something for a future PR :)

File: include/solvers/tracer_scratch_data.h [hunk 6]
```diff
@@ -260,6 +284,219 @@ class TracerScratchData
...
+    fe_interface_values_tracer.reinit(cell, face_no, sf, ncell, nf, nsf);
+    face_quadrature_points = fe_interface_values_tracer.get_quadrature_points();
+
+    n_interface_dofs = fe_interface_values_tracer.n_current_interface_dofs();
+
+    const double extent1 = cell->measure() / cell->face(face_no)->measure();
+    const double extent2 = ncell->measure() / ncell->face(nf)->measure();
+
+    penalty_factor =
+      get_penalty_factor(fe_values_tracer.get_fe().degree, extent1, extent2);
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_scratch_data.h:331]: It's a good idea, I modified it everywhere to remove extent1 and extent2

File: include/solvers/tracer_scratch_data.h [hunk 7]
```diff
@@ -260,6 +284,219 @@ class TracerScratchData
...
+          levelset_function != nullptr,
+          ExcMessage(
+            "Levelset function is required for tracer assembly, but the level set function is a nullptr"));
+
+        levelset_function->value_list(face_quadrature_points, sdf_face_values);
+      }
+  }
+
+  /** @brief Reinitializes the content of the scratch regarding the velocity for internal/boundary faces.
+   *  The velocity is inherently assumed to have been solved using a CG scheme.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_scratch_data.h:417]: Nope the comment is right. The velocity is assumed to be CG, meaning that there is only one velocity per face :)

File: include/solvers/tracer_scratch_data.h [hunk 8]
```diff
@@ -260,6 +284,219 @@ class TracerScratchData
...
+    const VectorType                                     &current_solution,
+    const Function<dim>                                  *levelset_function)
+  {
+    fe_interface_values_tracer.reinit(cell, face);
+    const FEFaceValuesBase<dim> &fe_face =
+      fe_interface_values_tracer.get_fe_face_values(0);
+    face_quadrature_points = fe_interface_values_tracer.get_quadrature_points();
+
+    // BB TODO: These could be pre-allocated
+    values_here.resize(face_quadrature_points.size());
```
[the remark this answers is not in the document]
Comment (oguevremont) [reply include/solvers/tracer_scratch_data.h:392]: You mean pre-allocated to the number of quadrature points for a face, instead of `n_q_points` (for a cell)?

File: source/solvers/tracer_assemblers.cc [hunk 9]
```diff
@@ -241,6 +241,102 @@ TracerAssemblerCore<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+
+  // Loop and quadrature informations
+  const auto        &JxW_vec    = scratch_data.JxW;
+  const unsigned int n_q_points = scratch_data.n_q_points;
+  const unsigned int n_dofs     = scratch_data.n_dofs;
+
+  // Copy data elements
+  auto &local_matrix = copy_data.local_matrix;
+
+  // assembling local matrix and right hand side
```
Comment (oguevremont) [inline source/solvers/tracer_assemblers.cc:262]:
```suggestion
  // Assembling local matrix
```

File: source/solvers/tracer_assemblers.cc [hunk 10]
```diff
@@ -241,6 +241,102 @@ TracerAssemblerCore<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+
+  // Loop and quadrature informations
+  const auto        &JxW_vec    = scratch_data.JxW;
+  const unsigned int n_q_points = scratch_data.n_q_points;
+  const unsigned int n_dofs     = scratch_data.n_dofs;
+
+  // Copy data elements
+  auto &local_rhs = copy_data.local_rhs;
+
+  // assembling local matrix and right hand side
```
Comment (oguevremont) [inline source/solvers/tracer_assemblers.cc:310]:
```suggestion
  // Assembling local right hand side
```

File: source/solvers/tracer_assemblers.cc [hunk 11]
```diff
@@ -345,3 +441,255 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+              if (boundary_conditions_tracer.type[boundary_index] ==
+                  BoundaryConditions::BoundaryType::outlet)
+                {
+                  if (velocity_dot_n > 0)
+                    copy_data.local_matrix(i, j) +=
+                      fe_face.shape_value(i, point) *
+                      fe_face.shape_value(j, point) * velocity_dot_n *
+                      JxW[point];
+                }
+              if (boundary_conditions_tracer.type[boundary_index] ==
```
Comment (oguevremont) [inline source/solvers/tracer_assemblers.cc:587]:
```suggestion
             else if (boundary_conditions_tracer.type[boundary_index] ==
```

File: include/solvers/tracer_assemblers.h [hunk 12]
```diff
@@ -22,11 +23,12 @@ class TracerAssemblerBase
 {
 public:
   /**
-   * @brief assemble_matrix Interface for the call to matrix assembly
-   * @param scratch_data Scratch data containing the Tracer information.
+   * @brief Interface for the call to matrix assembly.
+   * @param[in,out] scratch_data Scratch data containing the Tracer information.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_assemblers.h:27]: Good point, it should only be in everywhere. In fact the scratch data should be const also.  I'll fix that

File: source/solvers/tracer_assemblers.cc (hunk 9 above)
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:262]: I mostly mean pre-allocating the memory for the arrays instead of allocating them on the fly. When we re-allocate for every cell, we do pay a price for that. That will be done in a follow-up PR.

File: source/solvers/tracer_assemblers.cc (hunk 11 above)
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:587]: No this is not an else if here, this is really an if :).

Comment (oguevremont) [reply source/solvers/tracer_assemblers.cc:587]: Alright! A BC can be both outlet and dirichlet?

Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:587]: Ahh no, you are right, it's either one or the other.

File: include/solvers/tracer.h [hunk 13]
```diff
@@ -383,6 +423,120 @@ class Tracer : public AuxiliaryPhysics<dim, GlobalVectorType>
   void
   write_tracer_statistics();
 
+
+
+  /**
+   * @brief Get the lethe boundary indicator for a given triangulation boundary while carrying the appropriate checks
```
Comment (hepap) [inline include/solvers/tracer.h:429]:
```suggestion
   * @brief Get the Lethe boundary indicator for a given triangulation boundary while 
   * carrying the appropriate checks.
```

File: include/solvers/tracer_scratch_data.h [hunk 14]
```diff
@@ -260,6 +283,220 @@ class TracerScratchData
...
+   * @param[in] face_no The face index associated with the cell
+   *
+   * @param[in] sub_face_no The subface index associated with the face
+   *
+   * @param[in] ncell The neighboring cell
+   *
+   * @param[in] n_face_no The face index associated with the neighboring cell
+   *
+   * @param[in] n_sub_face_no The subface index associated with the neighboring
+   * cell
```
Comment (hepap) [inline include/solvers/tracer_scratch_data.h:300]: For readability (and it doesn't match the definition)
```suggestion
   * @param[in] neigh_cell The neighboring cell
   *
   * @param[in] neigh_face_no The face index associated with the neighboring cell
   *
   * @param[in] neigh_sub_face_no The subface index associated with the neighboring
   * cell
```

File: include/solvers/tracer_scratch_data.h [hunk 15]
```diff
@@ -260,6 +283,220 @@ class TracerScratchData
...
+          levelset_function != nullptr,
+          ExcMessage(
+            "Levelset function is required for tracer assembly, but the level set function is a nullptr"));
+
+        levelset_function->value_list(face_quadrature_points, sdf_face_values);
+      }
+  }
+
+
+  /** @brief Reinitializes the content of the scratch for the internal faces. This is only used for the DG assemblers.
```
Comment (hepap) [inline include/solvers/tracer_scratch_data.h:364]:
```suggestion
  /** @brief Reinitializes the content of the scratch for the boundary faces. This is only used for the DG assemblers.
```

**Review by hepap — COMMENTED** [review]
Here are my first comments, I still have the tests to look at. I will complete my review tomorrow :)

File: include/solvers/tracer_assemblers.h [hunk 16]
```diff
@@ -69,27 +72,72 @@ class TracerAssemblerCore : public TracerAssemblerBase<dim>
...
+   * @param[in] scratch_data (see base class)
+   * @param[in,out] copy_data (see base class)
+   */
+  virtual void
+  assemble_rhs(const TracerScratchData<dim> &scratch_data,
+               StabilizedMethodsCopyData    &copy_data) override;
+
+  // The simulation control must be a part of the core assembler since the
+  // inverse time-step is used within the stabilization
+  std::shared_ptr<SimulationControl> simulation_control;
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_assemblers.h:137]: Nope you are right. The mistake is putting it in the base class actually. will remove it from there and from where it's not needed (aka everywhere in the DG case)

File: source/solvers/tracer_assemblers.cc [hunk 17]
```diff
@@ -345,3 +446,256 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
 template class TracerAssemblerBDF<2>;
 template class TracerAssemblerBDF<3>;
+
+
+
+template <int dim>
+void
+TracerAssemblerSIPG<dim>::assemble_matrix(
+  const TracerScratchData<dim> &scratch_data,
+  StabilizedDGMethodsCopyData  &copy_data)
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:456]: Good question. I did not want to create a CopyData that was not stabilized for DG because initially I thought we would need stabilization. Since we don't, I can remove that.

File: source/solvers/tracer_assemblers.cc [hunk 18]
```diff
@@ -345,3 +446,256 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+        scratch_data.face_velocity_values[q] * normals[q];
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        for (unsigned int j = 0; j < n_dofs; ++j)
+          {
+            copy_data_face.face_matrix(i, j) +=
+              fe_iv.jump_in_shape_values(i, q) // [\phi_i]
+              * fe_iv.shape_value((velocity_dot_n > 0.),
+                                  j,
+                                  q) // phi_j^{upwind}
+              * velocity_dot_n       // (\beta .n)
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:480]: yeah that'S from where I took it, it should be u, y ou are right

File: source/solvers/tracer.cc [hunk 19]
```diff
@@ -80,6 +108,146 @@ Tracer<dim>::assemble_system_matrix()
...
+    };
+
+  const auto boundary_worker =
+    [&](const typename DoFHandler<dim>::active_cell_iterator &cell,
+        const unsigned int                                   &face_no,
+        TracerScratchData<dim>                               &scratch_data,
+        StabilizedDGMethodsCopyData                          &copy_data) {
+      // Identify which boundary condition corresponds to the boundary id. If
+      // this boundary condition is not identified, then exit the simulation
+      // instead of assuming an outlet.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer.cc:147]: That comment it outdated, I will fix that.

File: source/solvers/tracer.cc [hunk 20]
```diff
@@ -219,6 +400,138 @@ Tracer<dim>::assemble_system_rhs()
...
+        multiphysics->get_dof_handler(PhysicsID::fluid_dynamics);
+      // Identify the cell that corresponds to the fluid dynamics
+      typename DoFHandler<dim>::active_cell_iterator velocity_cell(
+        &(*triangulation), cell->level(), cell->index(), dof_handler_fluid);
+
+      reinit_face_velocity_with_adequate_solution(velocity_cell,
+                                                  face_no,
+                                                  scratch_data);
+
+      scratch_data.calculate_face_physical_properties();
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer.cc:460]: regrettably at the moment this would not be trivial. But I can try to fix that in the future. I don,t mind repeating 4 lines for the moment.

File: include/solvers/tracer_assemblers.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_assemblers.h:137]: It's not public in the base class actually, it's public in the core class... So you do need to repeat it...

File: source/solvers/tracer_assemblers.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:456]: Ohh I know why. If we do not make it inherit of the regular stabilized method copy data then we cannot use the regular cell assembly functions. So consequently, it means that we would need to rewrite the loop over local cells. This is an improvement that could be done in the future, but right now this has no cost and really reduces code duplication TBH.

File: source/solvers/tracer.cc (hunk 20 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer.cc]: This is also an issue we have with all RHS and Matrix assembly. Eventually what we could do would be to have a template argument to have the same function do RHS and matrix assembly. This would remove code duplication. I will open an issue on this.

File: include/solvers/copy_data.h [hunk 21]
```diff
@@ -114,6 +114,42 @@ class StabilizedMethodsCopyData
...
 };
 
+
+
+/**
+ * @brief Responsible for storing the information calculated using the discontinuous
+ * galerkin (DG) assembly of stabilized scalar equations. Like the
+ * StabilizedCopyData class, this class is used to initialize, zero (reset) and
+ * store the cell_matrix and the cell_rhs while having support for stabilized
+ * terms. However, it also has to support the storage of information at internal
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/copy_data.h:124]: The issue is taht i need the copy data to inherit from the Stabilized one in order to use the same functions.

File: include/solvers/tracer_assemblers.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (hepap) [reply include/solvers/tracer_assemblers.h:137]: Yeah you are right! Do you need it for DG assembler?

File: source/solvers/tracer_assemblers.cc [hunk 22]
```diff
@@ -345,3 +446,255 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+                * scratch_data.values_here[q]    // \phi_i^{upwind}
+                * velocity_dot_n                 // (\beta . n)
+                * JxW[q];                        // dx
+            }
+          else
+            {
+              copy_data_face.face_rhs(i) -=
+                fe_iv.jump_in_shape_values(i, q) // [\phi_i]
+                * scratch_data.values_there[q]   // \phi_i^{upwind}
+                * velocity_dot_n                 // (\beta . n)
```
Comment (hepap) [inline source/solvers/tracer_assemblers.cc:535]:
```suggestion
                * velocity_dot_n                 // (u . n)
```

File: source/solvers/tracer_assemblers.cc [hunk 23]
```diff
@@ -345,3 +446,255 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+    {
+      for (unsigned int point = 0; point < q_points.size(); ++point)
+        {
+          const double velocity_dot_n =
+            scratch_data.face_velocity_values[point] * normals[point];
+
+          for (unsigned int i = 0; i < n_facet_dofs; ++i)
+            copy_data.local_rhs(i) -= fe_face.shape_value(i, point) // \phi_i
+                                      * scratch_data.values_here[point] *
+                                      velocity_dot_n // \beta . n
```
Comment (hepap) [inline source/solvers/tracer_assemblers.cc:647]:
```suggestion
                                      velocity_dot_n // u . n
```

File: applications_tests/lethe-fluid/dg_tracer_diffusion_mms_2d.output [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+cells  error_velocity   error_pressure  
+    4 0.000000e+00   - 0.000000e+00   - 
+   16 0.000000e+00 nan 0.000000e+00 nan 
+   64 0.000000e+00 nan 0.000000e+00 nan 
+  256 0.000000e+00 nan 0.000000e+00 nan 
+cells   error_tracer    
+    4 4.161920e-03    - 
+   16 5.293724e-04 2.97 
+   64 3.916158e-05 3.76 
+  256 2.547542e-06 3.94
```
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-fluid/dg_tracer_diffusion_mms_2d.output:115]: I think here it's an artifact of DG....

File: source/solvers/tracer_assemblers.cc (hunk 9 above)
Comment (oguevremont) [reply source/solvers/tracer_assemblers.cc:264]: I see!

**Review by oguevremont — APPROVED** [review]
Really nice addition

File: applications_tests/lethe-fluid/dg_tracer_circular_advection.mpirun=2.output [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+   Number of active cells:       1024
+   Number of degrees of freedom: 3267
+   Volume of triangulation:      1
+   Number of tracer degrees of freedom: 4096
+L2 error velocity : 1.63299
+L2 error tracer : 2.77243e-05
+cells   error_velocity    error_pressure  
+   64 1.632993e+00     - 0.000000e+00   - 
+  256 1.632993e+00 0.00 0.000000e+00 nan 
+ 1024 1.632993e+00  0.00 0.000000e+00 nan
```
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-fluid/dg_tracer_circular_advection.mpirun=2.output:35]: It does, but I guess that's life :)

File: source/solvers/tracer_assemblers.cc [hunk 26]
```diff
@@ -345,3 +446,255 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+        scratch_data.face_velocity_values[q] * normals[q];
+      for (unsigned int i = 0; i < n_dofs; ++i)
+        {
+          // Assemble advection terms with upwinding
+          if (velocity_dot_n > 0)
+            {
+              copy_data_face.face_rhs(i) -=
+                fe_iv.jump_in_shape_values(i, q) // [\phi_i]
+                * scratch_data.values_here[q]    // \phi_i^{upwind}
+                * velocity_dot_n                 // (\beta . n)
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/tracer_assemblers.cc:527]: We did not need it. I removed it everywhere
